download/download.go

```diff
@@ -72,35 +73,68 @@ func (d Downloader) Download(ctx context.Context, pair, timeframe string, output
 
 	parameters.Start = time.Date(parameters.Start.Year(), parameters.Start.Month(), parameters.Start.Day(),
 		0, 0, 0, 0, time.UTC)
-	parameters.End = time.Date(parameters.End.Year(), parameters.End.Month(), parameters.End.Day(),
-		0, 0, 0, 0, time.UTC)
+
+	if now.Sub(parameters.End) > 0 {
+		parameters.End = time.Date(parameters.End.Year(), parameters.End.Month(), parameters.End.Day(),
+			0, 0, 0, 0, time.UTC)
+	} else {
+		parameters.End = now
+	}
 
 	candlesCount, interval, err := candlesCount(parameters.Start, parameters.End, timeframe)
 	if err != nil {
 		return err
 	}
+	candlesCount++
 
 	log.Infof("Downloading %d candles of %s for %s", candlesCount, timeframe, pair)
 	info := d.exchange.AssetsInfo(pair)
 	writer := csv.NewWriter(recordFile)
+
+	progressBar := progressbar.Default(int64(candlesCount))
+	lostData := 0
+	isLastLoop := false
+
 	for begin := parameters.Start; begin.Before(parameters.End); begin = begin.Add(interval * batchSize) {
 		end := begin.Add(interval * batchSize)
 		if end.After(parameters.End) {
 			end = parameters.End
+			isLastLoop = true
+		} else {
+			end = end.Add(-1 * time.Second)
```

Comment:
It is used to avoid the duplication of the last candle?

Reply:
yes, Binance api (kLine History) use nano unix to query add or minus 1 second  ,it can avoid duplicate between last candle and the first candle of next loop